src/codegen/exeffToNoeff.ml

```diff
@@ -0,0 +1,360 @@
+              ( (var, t1, t2, (pattern_elab p, elabc)) ) ) in
+      let (tycomp, elabcomp) = comp_elab state (extend_env env abs_list) comp in
+      ( tycomp, NoEff.NLetRec (List.map (elab_letrec_abs) abs_list, elabcomp) )
+  | ExEff.Match (value, abs_lst, loc) ->
+    let (tyv, elabv) = value_elab state env value in
+    let elab_abs (pat, comp) = 
+            ( let (tyc, elabc) = comp_elab state env comp in (pattern_elab pat, elabc) ) in
+    ( if ( (List.length abs_lst) = 0)
+    then ( failwith "Empty match statement" )
+    else ( let ((p1,c1) :: _ ) = abs_lst in
+           let (tyc, elabc) = comp_elab state env c1 in 
+           (tyc, NoEff.NMatch (elabv, List.map elab_abs abs_lst, loc)) ) )
+  | ExEff.Apply (v1, v2) -> 
+    let (ty1, elab1) = value_elab state env v1 in
+    ( match ty1 with
+      | ExEffTypes.Arrow (t1, t2) -> 
+        let (ty2, elab2) = value_elab state env v2 in
+          if (ty2 = t1)
+          then (t2, NoEff.NApplyTerm (elab1, elab2))
+          else (failwith "Improper argument type")
+      | _ -> failwith "Improper function type" )
+  | ExEff.Handle (value, comp) -> 
+    let ( (ctype, cdirt), elabc ) = comp_elab state env comp in 
+    let ( vtype, velab ) = value_elab state env value in
+    ( match vtype with
+      | ExEffTypes.Handler ( (vty1, vdirt1), (vty2, vdirt2) ) ->
+        if (vty1 = ctype && vdirt1 = cdirt) then (
+          if (Types.is_empty_dirt cdirt)
+          (* Handle - Case 1 *)
+          then ( (vty2, vdirt2), NoEff.NApplyTerm (velab, elabc))
+          else (
+               if (Types.is_empty_dirt vdirt2)
+               (* Handle - Case 2 *)
+               then ( let (_, telab) = type_elab state env vty2 in
+                 ( (vty2, vdirt2), 
+                 NoEff.NCast (NoEff.NHandle (elabc, velab),
+                 NoEff.NCoerUnsafe (NoEff.NCoerRefl (telab)))) )
+               (* Handle - Case 3 *)
+               else ( (vty2, vdirt2), NoEff.NHandle (elabc, velab))
+               )
+        )
+        else failwith "Handler source type and handled computation type do not match"
+      | _ -> failwith "Ill-typed handler" )
+  | ExEff.Call ((eff, (ty1, ty2)), value, (p, ty, comp)) ->
+    let (_, t1) = type_elab state env ty1 in
+    let (_, t2) = type_elab state env ty2 in
+    let (_, tt) = type_elab state env ty in
+    let (vty, velab) = value_elab state env value in
+    if (vty = ty1) 
+    then (
+      let env' = extend_pattern_type env p ty in
+      let (cty, celab) = comp_elab state env' comp in
+      ( cty, NoEff.NCall ((eff, (t1, t2)), velab, (pattern_elab p, tt, celab)) )
+    )
+    else failwith "Ill-typed call"
+  | ExEff.Op ((eff, (ty1, ty2)), value) -> 
+    let (_, t1) = type_elab state env ty1 in
+    let (_, t2) = type_elab state env ty2 in
+    let (vty, velab) = value_elab state env value in    
+    if (vty = ty1)
+    then ( ((ty2, ExEffTypes.empty_dirt), NoEff.NOp ((eff, (t1,t2)), velab)) )
+    else failwith "Ill-typed operation"
```

Comment:
We might have to change this but for now let's not bother. I still have to see what's the use of `Op` in the compiler (it looks as if it is never created in `explicitInfer`).